proxy/coco/src/request/states.rs

```diff
@@ -0,0 +1,308 @@
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Clones(usize);
+
+impl Clones {
+    /// Create a new `Clones` wrapping around `n`.
+    #[must_use]
+    pub const fn new(n: usize) -> Self {
+        Self(n)
+    }
+}
+
+impl From<Clones> for usize {
+    fn from(other: Clones) -> Self {
+        other.0
+    }
+}
+
+impl Add<usize> for Clones {
+    type Output = Self;
+
+    fn add(self, other: usize) -> Self::Output {
+        Self(self.0 + other)
+    }
+}
+
+impl AddAssign<usize> for Clones {
+    fn add_assign(&mut self, other: usize) {
+        *self = Self(self.0 + other)
+    }
+}
+
+/// The number of different attempts a `Request` has made during its lifetime.
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Attempts {
+    /// The number of query attempts we have made.
+    pub(super) queries: Queries,
+    /// The number of clone attempts we have made.
+    pub(super) clones: Clones,
+}
+
+impl Attempts {
+    /// Get a new `Attempts` where the number of queires and clones is initially `0`.
+    #[must_use]
+    pub const fn new() -> Self {
+        Attempts {
+            queries: Queries(0),
+            clones: Clones(0),
+        }
+    }
+}
+
+impl Default for Attempts {
+    fn default() -> Self {
+        Attempts::new()
+    }
+}
+
+impl sealed::Sealed for Attempts {}
+
+// State Traits
+
+/// If a state type implements this trait then it means that the type is allowed to increment the
+/// query attempts in a `Request`.
+///
+/// The trait is sealed internally, so we do not expect end-users to implement it.
+pub trait QueryAttempt: sealed::Sealed {}
+impl QueryAttempt for IsRequested {}
+
+/// If a state type implements this trait then it means that the type holds a `HashMap` of peers and
+/// their status of cloning.
+///
+/// The trait is sealed internally, so we do not expect end-users to implement it.
+pub trait HasPeers: sealed::Sealed
+where
+    Self: Sized,
+{
+    /// Give back the underlying `HashMap` of peers that is contained in `Self`.
+    fn peers(&mut self) -> &mut HashMap<PeerId, Status>;
+
+    /// Returns `false` if the `peers` are empty or if any of them are `Status::Available`
+    /// or `Status::InProgress`.
+    ///
+    /// Otherwise, if all are in the `Status::Failed` state, then we return `true`.
+    fn all_failed(&mut self) -> bool {
+        let peers = self.peers();
+
+        if peers.is_empty() {
+            return false;
+        }
+
+        peers.iter().all(|(_, status)| *status == Status::Failed)
+    }
+}
+
+impl HasPeers for Found {
+    fn peers(&mut self) -> &mut HashMap<PeerId, Status> {
+        &mut self.peers
+    }
+}
+
+impl HasPeers for Cloning {
+    fn peers(&mut self) -> &mut HashMap<PeerId, Status> {
+        &mut self.peers
+    }
+}
+
+/// If a state type implements this trait it means that their is a valid transition from that state
```

Comment:
Did you mean "this"?
```suggestion
/// If a state type implements this trait it means that this is a valid transition from that state
```

Reply:
I meant "there". How embarrassing :flushed: